### PR TITLE
[Ready] Json-fy  Mutation Categories

### PR DESCRIPTION
**Pull request by Frost-wood**
I was looking on the dwarf fortress forums and saw a post that looked to be an interesting idea for a mutation mod.  Upon trying to create a mod, I discovered a problem-when trying to make a targetable mutagen for a new category, and then discovered that you cannot actually create a targetable mutagen or iv for a specific category, without touching the code.  I tried using a lua script, but it seems the lua is not robust enough for that(or my understanding of lua is not enough).

This PR aims to fix that by allowing modders to create their own Mutation Category, while moving the core mutation categories to json.

Right now every mutation category for the weak-mutagen seems to work fine being moved to json, so all that is left is moving the iv mutation handling to json and updating the jsons.

### CODE REVIEW

**Comment by kevingranade** [comment]
Cool, looks like we're having another wave of jsonification.

**Comment by DavidKeaton** [comment]
very nice!

File: src/iuse.cpp [hunk 1]
```diff
@@ -1673,74 +1673,20 @@ int iuse::mutagen(player *p, item *it, bool, point)
...
-        p->mutate_category(mutation_category);
-        p->mod_pain(2 * rng(1, 5));
-        p->hunger += 10;
-        p->fatigue += 5;
-        p->thirst += 10;
+        for (auto& iter : mutation_category_traits){
+            mutation_category_trait m_category = iter.second;
+            if (it->has_flag("MUTAGEN_" + m_category.category)) {
+                mutation_category = "MUTCAT_" + m_category.category;
+                p->add_msg_if_player(_(m_category.mutagen_message.c_str()));
```
Comment (BevapDin) [inline src/iuse.cpp]: The message is already translated (the `_` does it) when the category is loaded. No need to do this here again.

File: data/json/mutation_category.json [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		"iv_message" : "",		
+		"memorial_message" : "NULL"		
+	},{
+		"type" : "mutation_category",
+		"id" : "mutation_raptor",
+		"name" : "Raptor",
+		"category" : "RAPTOR",
+		"mutagen_message" : "Mmm...sweet, bloody flavor...tastes like victory.",
+		"iv_message" : "",		
+		"memorial_message" : "NULL"
```
Comment (BevapDin) [inline data/json/mutation_category.json]: If the message entries are optional anyway, just remove them here and add the default value when loading them. That makes the json simpler.

```C++
new_category.iv_message = jsobj.get_string( "iv_message", "" ); // second parameter is the default value
```

The translation system will translate the empty string into some debug information about the translation data, you probably don't want that. Check whether the message is empty before translating it. You can see the translation of the empty string in #10882.
```C++
if( !new_category.iv_message.empty() ) {
    new_category.iv_message = _(new_category.iv_message.c_str());
}
```

Comment (Frost-wood) [inline data/json/mutation_category.json]: I haven't worked my way to it yet, but the strings aren't optional, I tested it, and the game will throw a json error if someone tries to create a category without these base strings.

File: src/iuse.cpp (hunk 1 above)
Comment (Frost-wood) [inline src/iuse.cpp]: So the _ is part of the translation system... I will keep that in mind.

File: data/json/mutation_category.json (hunk 2 above)
Comment (narc0tiq) [inline data/json/mutation_category.json]: > throw a json error

Well, sure, if you don't set a default value, the value is then required. It's your code and your call to `json_object.get_string()`.

File: src/mutation_data.cpp [hunk 3]
```diff
@@ -35,7 +36,43 @@ static void load_mutation_mods(JsonObject &jsobj, std::string member, std::unord
...
+    new_category.iv_fatigue  = jsobj.get_int("iv_fatigue", 5);
+    new_category.iv_morale   = jsobj.get_int("iv_morale", 0);
+    new_category.iv_morale_max   = jsobj.get_int("iv_morale_max", 0);
+    new_category.iv_sound = jsobj.get_bool("iv_sound", false);
+    new_category.iv_sound_message = jsobj.get_string(("iv_sound_message"), "NULL");
+    new_category.iv_noise = jsobj.get_int("iv_noise", 0);
+    new_category.iv_sleep = jsobj.get_bool("iv_sleep", false);
+    new_category.iv_sleep_message =_(jsobj.get_string("iv_sleep_message", "NULL").c_str());
+    new_category.iv_sleep_dur = jsobj.get_int("iv_sleep_dur", 0);
+    new_category.memorial_message = _(jsobj.get_string("memorial_message").c_str());
```
Comment (narc0tiq) [inline src/mutation_data.cpp]: Re: [previous comment](https://github.com/CleverRaven/Cataclysm-DDA/pull/11522/files#r26188306) -- see, right here, you have some strings being fetched from json with defaults above, and here's `memorial_message`, right in your diff. Just like @BevapDin said, if you give it a default value, you can omit it from the JSON.

**Comment by Frost-wood** [comment]
Hopefully its clear why I didn't put defaults in, however if people still feel that defaults are required after looking at the finished code, then I'll put them in.

File: src/player.cpp [hunk 4]
```diff
@@ -95,8 +95,7 @@ void game::init_morale()
     _("Killed Friend"),
     _("Guilty about Killing"),
     _("Guilty about Mutilating Corpse"),
-    _("Chimerical Mutation"),
-    _("Fey Mutation"),
+    _("Mutation"),
```
Comment (KA101) [inline src/player.cpp]: Not thrilled at losing the specific rationale, but wev.

File: src/pldata.h [hunk 5]
```diff
@@ -169,6 +169,39 @@ class addiction : public JsonSerializer, public JsonDeserializer
...
+    int iv_hunger   = 10;
+    int iv_thirst   = 10;
+    int iv_pain     = 2;
+    int iv_fatigue  = 5;
+    int iv_morale   = 0;
+    int iv_morale_max = 0;
+    bool iv_sound = false;  //determines if you make a sound when you inject mutagen
+    std::string iv_sound_message = "NULL";
+    int iv_noise = 0;    //the amount of noise produced by the sound
+    bool iv_sleep = false;  //whether the iv has a chance of putting you to sleep
```
Comment (KA101) [inline src/pldata.h]: Not so much "sleep" as short-term KO from the process.

File: src/player.cpp (hunk 4 above)
Comment (Frost-wood) [inline src/player.cpp]: I can keep them if you are alright with the added complexity in a if statement.  I'm not sure why add_morale isn't designed to accept strings(performance reasons?)

Comment (KA101) [inline src/player.cpp]: I don't mind if statements, but then I'm not opposed to mutations being C++; this still needs C++ to handle effects, anyway.

**Comment by KA101** [comment]
Hey, heard a report that apparently mutations/traits aren't being preserved in saves.  Didn't think this would break those and didn't specifically check, but since it's a recent merge, thought I'd mention it.  Anyone had issues with traits/mutations in saves?

**Comment by Frost-wood** [comment]
I'll check

**Comment by Frost-wood** [comment]
Everything seems to work fine(templates load fine, and mutations are preserved), if it is the report that I think you are talking about it looks like the person in question modified the template manually.

**Pull request merged**